File: Cpp/1695_maximum_erasure_value.cpp
// https://leetcode.cn/problems/maximum-erasure-value/

#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int res = 0;
        int left = 0, tmp = 0;
        unordered_set<int> s;
        for (int i = 0; i < nums.size(); ++i) {
            int num = nums[i];
            while (left < i && s.find(num) != s.end()) {
                tmp -= nums[left];
                s.erase(nums[left]);
                left += 1;
            }
            tmp += num;
            s.insert(num);
            res = max(res, tmp);
        }
        return res;
    }
};

int main() {
    Solution sol;
    vector<int> nums = {4,2,4,5,6};
    cout << sol.maximumUniqueSubarray(nums) << '\n';
    return 0;
}

File: Python3/1695_maximum_erasure_value.py
# https://leetcode.cn/problems/maximum-erasure-value/

class Solution:
    def maximumUniqueSubarray(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: int
        """
        res = 0
        left, tmp = 0, 0
        s = set()
        for i, num in enumerate(nums):
            while left < i and num in s:
                tmp -= nums[left]
                s.remove(nums[left])
                left += 1
            tmp += num
            s.add(num)
            res = max(res, tmp)
        return res